src/printer2/index.ts

```diff
@@ -0,0 +1,61 @@
+import { RootNode } from "../ast/rootNode";
+import {
+  Printer,
+  lastOrUndefined,
+  printDetachedComments,
+  printLeadingComments,
+} from "./common";
+import { toString, concat } from "./formatter";
+
+/**
+ * Create a new printer instance.
+ */
+function createPrinter(source: string) {
+  return {
+    source,
+    containerPos: 0,
+    containerEnd: 0,
+    declarationListContainerEnd: 0,
+    lineMap: [],
+    detachedCommentsInfo: undefined,
+  };
+}
+
+export function printSource(root: RootNode): any {
```

Comment:
@aladdin-add What do you think? Too expensive? I had to build a container around the 'RootNode' and the same around statements, expressions and types...